### DISCUSSION TITLE
Quicksync (QSV) scaling with FFMPEG runOnDemand

### DISCUSSION THREAD

**Discussion by clee604** [bluenviron/mediamtx#2929] [Questions] [answered]
### Question

This question likely is more related to ffmpeg than mediamtx, but I'm hoping someone way smarter than me can help out and that this is just a trival missing parameter.

I have a runOnDemand command that sources a rtsp feed from a wyze camera and I'm looking to use quicksync with ffmpeg to downscale the video to 720p from the 1080p feed.

Using this command, quicksync works fine:
runOnDemand: ffmpeg -acodec pcm_s16le -hwaccel qsv -rtsp_transport tcp -i rtsp://localhost:8554/wyze -c:v h264_qsv -bf 0 -vf scale=1280:720 -c:a aac -f rtsp rtsp://localhost:$RTSP_PORT/$RTSP_PATH

However I learned this uses a software scaler, and that QSV actually has a hw scaler 'scale_qsv'

Modifying the command to use this, however results in an error:
runOnDemand: ffmpeg -acodec pcm_s16le -hwaccel qsv -hwaccel_output_format qsv -rtsp_transport tcp -i rtsp://localhost:8554/wyze -c:v h264_qsv -bf 0 -vf "scale_qsv=w=1280:h=720:" -c:a aac -f rtsp rtsp://localhost:$RTSP_PORT/$RTSP_PATH

Impossible to convert between the formats supported by the filter 'graph 0 input from stream 0:0' and the filter 'auto_scaler_0'

I've tried combinations of adding mode=hq,hwmap=derive_device=qsv,hwdownload to the parameters with the same result. Am I doing something impossible or stupid?

**Comment by dragulesq** [comment, marked as answer]
I think I went through the same painful trial and error process and probably saw that error too. I was using ffmpeg 6.1. Here is what works for me but I compiled ffmpeg from source ( ffmpeg version N-112753-gce467421dc) because there was a commit that fixed some QSV issues around the time I was testing. In the example below I am also reducing the framerate for yolo inference and tracking.
```
runOnReady: >
      /usr/local/bin/ffmpeg -hide_banner -nostats -loglevel warning -vcodec h264_qsv -hwaccel qsv -hwaccel_output_format qsv
        -rtsp_transport tcp -i rtsp://localhost:$RTSP_PORT/pi3 -vf "hwupload,vpp_qsv=framerate=12,scale_qsv=w=-1:h=640"
        -framerate 12 -b:v 1500k -vcodec h264_qsv -preset slow -max_muxing_queue_size 1024 -g 12 -f rtsp rtsp://localhost:$RTSP_PORT/pi3lo
```

Let me know if it works and what version of ffmpeg you are running. Good luck!

**Reply by clee604** [reply]
Looks like I'm using an (very?) old version of ffmpeg... 
ffmpeg -version
ffmpeg version 4.3.6-0+deb11u1 Copyright (c) 2000-2023 the FFmpeg developers
built with gcc 10 (Debian 10.2.1-6)

I tried downloading a precompiled binary but it looks like that one doesn't have the h264_qsv encoder

Guess my first order of business is to figure out how to update ffmpeg with the right packages

**Reply by dragulesq** [reply]
Try first a static build from here https://johnvansickle.com/ffmpeg/
Maybe the master built on 20240112

**Reply by clee604** [reply]
Thanks! Had to go down the rabbit hole of updating my LXC container to Bookworm (which updated ffmpeg to version 5.1)... still didn't work.. then I updated to ffmpeg 6.1 using deb-multimedia and now it works! So the key factor was probably having to use ffmpeg 6.1.

Did you do any testing with the presets or other parameters you used? Wondering if I could tune it a bit better since the quality seems considerably worse than software transcoding and latency is actually slightly worse.

**Reply by dragulesq** [reply]
Awesome! Yea you have to experiment with the settings depending on your network/hardware and quality requirements. I am okay with a delay actually since I am doing detection and tracking. I am on a wireless network and had to downgrade the bitrate. You could try UDP  and the low latency flags (nobuffer, smaller probesize etc)  if you want  better latency.

[a comment by an automated account is not shown]
